### PR TITLE
Endpoints

### PR DESCRIPTION
**Pull request by Dunn-Austen**
**This portion of the PR refers to our re-designed project**
Summary: We decided to abandon our users table and alter our projected endpoints and table setups. This PR now suffices to reset our migration and seed data to fit the new project designs. It wipes out obsolete code and replaces it with the updated foundation for the new code base.

**Instructions**
- Make sure you've rolled back your old migration code before merging this down
- Once reviewed and merged, you may have to install SuperHost, Babel, & Jest (see background information in former section of the PR)
- Run the following commands to get your local DB up to speed:
```
knex migrate: latest
knex migrate: --env=test
knex seed:run
knex seed:run --env=test
```

___
**The text below refers to edits made before fundamentally altering our tables (removing users)**
### What does this PR do?

Summary: Most of these edits are about bringing our codebase up to speed so that we can write endpoints and test them. Lots of configuration tweaking, installs, changes to the knexfile, package.json, etc. After this PR, you should be able to freely write tests. If not, then there are a couple places of friction that we can investigate. First, the migration files might cause you trouble. I'll walk you through that. Second, you may have to run some installs on your local machine: Jest, Babel, Supertest, etc. See the bottom section (background info) for specific detail.

In detail:
-  Splits the server.js code into a newly created app.js file
- Adds a seed property to our knexfile object
- Adds a test property to our knexfile object
- Adds a Dev directory to our seeds folder, and adjusts file locations & imports accordingly
- Migrates the seeds folder to within the Migrations directory
- Installs and configures Babel - including adjustments to package.json  (allowing for use of IMPORT syntax from ES6 - as opposed to REQUIRE)
- Installs and configures Jest - including adjustments to package.json  (allowing for use of async/await in testing)
- Builds out the Test directory within Seeds
- Adds a user.js file with a couple default users
- Updates the seed files so that they populate the users table with our default users
- Adds a second migration which amends the columns for users (adding password, and removing created_at and updated_at)
- Adds code to exports.seed so that the Primary ID resets (I was running into issues with the incrementing starting higher than 0)
- Adds a happy path test for GET - a single user
- Adds the endpoint for GET - a single user
- Installs and configures Supertest 
- Adds default Projects objects (projects.js)
- Accounts for projects code in the seed files (dev and test)

### Where should the reviewer start?
- Follow the changed files in what should be a meticulous review due to the large volume of configuration edits. It needs to all make sense before being merged.

### Any background information?
- I may have used an older version of Babel, but it seems to work fine for now. If we need to change it, we would also have to update the Babel imports in app.js and app.test.js
This command should get you up to speed. I used this old Turing lesson to bridge the gaps for things that were not explicitly broken down in the lessons from this mod.

Lesson: https://frontend.turing.io/lessons/module-3/server-side-testing.html

```
npm install @babel/polyfill babel-jest supertest -D
```
- Migration issues. So I made some mistakes that might require some joint action. In trying to add a migration that would remove the palette_ids, created_at, and updated_at columns while adding a password column to the Users table - made the mistake of rolling back the second migration. I realized after adding the password column that I'd wanted to drop some of the others, and figured a rollback would be okay in this situation. Unfortunately, the rollback code wiped all the tables. This created a complication between the initial and second migration. You'll notice in the second migration file that I have basically all the code from the first. Ideally the second migration would have very little code accounting for just the precise edits. Since I'd removed the tables entirely, I couldn't use dropColumn in isolation. I had to repeat the createTable actions. 
This caused additional complexity once I'd built out the Test database - since it's subject to the same migrations for table construction. I had to do some strange stuff to get it all to work. **My recommendation is that we rollback all the migrations due to the mess I made and start from scratch with the appropriate headings.** Things works now for me, but will be strange for you. 
- Do we need to use promises in the seed code for Projects (since it relies on user_id from the users table)?

### CODE REVIEW

File: app.test.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  });
+
+  describe('init', () => {
+    it('should return a 200 status', async () => {
+      const res = await request(app).get('/')
+      expect(res.status).toBe(200)
+    });
+  });
+
+  describe('GET /api/v1/users/:id', () => {
```
Comment (Dustin-Mikusko) [inline app.test.js]: Same comment as in `app.js`, this methodology should actually be a POST request.

File: db/migrations/20200203183040_initial.js [hunk 2]
```diff
@@ -28,10 +28,10 @@ exports.up = function(knex) {
       table.integer('color_2_id').unsigned();
```
Comment (Dustin-Mikusko) [inline db/migrations/20200203183040_initial.js]: I'm still skeptical if this is the best way to go about this. Your thoughts on having the projects holding color IDs, or just storing the hex code correctly? We can discuss this more in detail in person, but I'm not sure I'm married to this initial setup of mine...

File: db/migrations/20200204174836_adjust-users-columns.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
+
+exports.up = function(knex) {
+  return knex.schema
+  .createTable('users', table => {
+    table.increments('id').primary();
+    table.string('email');
+    table.string('password');
```
Comment (Dustin-Mikusko) [inline db/migrations/20200204174836_adjust-users-columns.js]: Much needed. Thank you for catching this

File: db/seeds/dev/colors.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  try {
+    await knex('colors').del();
+    await knex.raw('ALTER SEQUENCE colors_id_seq RESTART WITH 1');
+    await knex('users').del();
+    await knex.raw('ALTER SEQUENCE users_id_seq RESTART WITH 1');
+    await knex('projects').del();
+    await knex.raw('ALTER SEQUENCE users_id_seq RESTART WITH 1');
+    await knex('projects').insert(theUsers);
+    await knex('users').insert(theUsers);
+
```
Comment (Dustin-Mikusko) [inline db/seeds/dev/colors.js]: This is all new to me. Is this effectively making sure that the IDs start at 1 each time a seed is run?

File: projects.js [hunk 5]
```diff
@@ -0,0 +1,20 @@
+const users = [
```
Comment (Dustin-Mikusko) [inline projects.js]: These are just dummy PROJECTS, not users, right?

File: db/seeds/test/colors.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  try {
+    await knex('colors').del();
+    await knex.raw('ALTER SEQUENCE colors_id_seq RESTART WITH 1');
+    await knex('users').del();
+    await knex.raw('ALTER SEQUENCE users_id_seq RESTART WITH 1');
+    await knex('projects').del();
+    await knex.raw('ALTER SEQUENCE users_id_seq RESTART WITH 1');
+    await knex('projects').insert(theUsers);
+    await knex('users').insert(theUsers);
+
```
Comment (Dustin-Mikusko) [inline db/seeds/test/colors.js]: This is just to ensure that each id starts at 1 upon a new seeding right?

File: db/seeds/dev/colors.js (hunk 4 above)
Comment (Dunn-Austen) [reply db/seeds/dev/colors.js]: I ran into some complications after a migration rollback. I re-migrated, re-seeded, and saw that the starting ID's were off - almost as if there were phantom values remaining from the previous seed data in table. I tried truncating the tables and reseeding, yet still ran into starting IDs that were non-zero.- which caused issues with matching ID's in my GET endpoint testing.

This code forces the column IDs to start at 1

File: projects.js (hunk 5 above)
Comment (Dunn-Austen) [reply projects.js]: Good catch. Thanks, man

File: db/migrations/20200203183040_initial.js [hunk 7]
```diff
@@ -17,34 +13,25 @@ exports.up = function(knex) {
...
-        .references('colors.id');
+      table.string('title');
+      table.string('color_1_id');
+      table.string('color_2_id');
+      table.string('color_3_id');
+      table.string('color_4_id');
+      table.string('color_5_id');
+      table.integer('project_id');
+      table.foreign('project_id')
+        .references('projects.id');
```
Comment (Dustin-Mikusko) [inline db/migrations/20200203183040_initial.js]: This is way cleaner, and frankly, simpler

**Pull request merged**